File: markdowner-core/src/test/java/org/jesperancinha/parser/markdowner/ReadmeNamingParserTest.java
package org.jesperancinha.parser.markdowner;

import org.apache.commons.io.IOUtils;
import org.jesperancinha.parser.markdowner.filter.FileFilterChain;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.nio.charset.Charset;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;

public class ReadmeNamingParserTest {

    private Path templatePath = Path.of("/.");

    @Test
    public void testBuildReadmeSamePath() throws IOException {
        final Path path = Path.of("/.");

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNull();
    }

    @Test
    public void testBuildReadmeStreamNothing() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/noProject2").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNull();
    }

    @Test
    public void testBuildReadmeStreamMixMavenAndNPMFlagOn() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project3MavenAndNPM").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().isNoEmpty(true).templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNull();
    }

    @Test
    public void testBuildReadmeStreamMixMavenAndNPM() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project3MavenAndNPM").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# This is a test project");
    }

    @Test
    public void testBuildReadmeStreamMavenName() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project1Maven").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# This is a test project");
    }

    @Test
    public void testBuildReadmeStreamMavenNoName() throws URISyntaxException, IOException {

        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project1MavenNoName").toURI());
        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# ProjectMavenArtifact");
    }

    @Test
    public void testBuildReadmeStreamNPM() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project2NPM").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# npm-project");
    }

    @Test
    public void testBuildReadmeStreamGradle() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project4Gradle").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# project4Gradle");
    }

    @Test
    public void testBuildReadmeStreamSBT() throws URISyntaxException, IOException {
        final Path path = Path.of(getClass().getResource("/directory2NoReadme/project5Sbt").toURI());

        final InputStream inputStream = ReadmeNamingParser.builder().templateLocation(templatePath).fileFilterChain(FileFilterChain.createDefaultChain()).build().buildReadmeStream(path);

        assertThat(inputStream).isNotNull();
        final String result = IOUtils.toString(inputStream, Charset.defaultCharset());
        assertThat(result).isEqualTo("# sbt-project");
    }

}

File: markdowner-filter/src/test/kotlin/org/jesperancinha/parser/markdowner/filter/ReadmeNamingParserTest.kt
package org.jesperancinha.parser.markdowner.filter

import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import org.apache.commons.io.IOUtils
import org.junit.jupiter.api.Test
import java.io.IOException
import java.net.URISyntaxException
import java.nio.charset.Charset.*
import java.nio.file.Path

class ReadmeNamingParserTest {
    private val templatePath = Path.of("/.")

    @Test
    @Throws(IOException::class)
    fun testBuildReadmeSamePath() {
        val path = Path.of("/.")
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        )
            .buildReadmeStream(path)
        inputStream.shouldBeNull()
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamNothing() {
        val noProject = javaClass.getResource("/directory2NoReadme/noProject2")
        noProject.shouldNotBeNull()
        val path = Path.of(noProject.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldBeNull()
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamMixMavenAndNPMFlagOn() {
        val projectMavenAndNpm = javaClass.getResource("/directory2NoReadme/project3MavenAndNPM")
        projectMavenAndNpm.shouldNotBeNull()
        val path = Path.of(projectMavenAndNpm.toURI())
        val inputStream = ReadmeNamingParser(
            isNoEmpty = true,
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldBeNull()
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamMixMavenAndNPM() {
        val projectMavenAndNpm = javaClass.getResource("/directory2NoReadme/project3MavenAndNPM")
        projectMavenAndNpm.shouldNotBeNull()
        val path = Path.of(projectMavenAndNpm.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# This is a test project"
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamMavenName() {
        val projectMaven = javaClass.getResource("/directory2NoReadme/project1Maven")
        projectMaven.shouldNotBeNull()
        val path = Path.of(projectMaven.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# This is a test project"
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamMavenNoName() {
        val projectMavenNoName = javaClass.getResource("/directory2NoReadme/project1MavenNoName")
        val path = Path.of(projectMavenNoName.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# ProjectMavenArtifact"
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamNPM() {
        val projectNpm = javaClass.getResource("/directory2NoReadme/project2NPM")
        projectNpm.shouldNotBeNull()
        val path = Path.of(projectNpm.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.Companion.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# npm-project"
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamGradle() {
        val projectGradle = javaClass.getResource("/directory2NoReadme/project4Gradle")
        projectGradle.shouldNotBeNull()
        val path = Path.of(projectGradle.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# project4Gradle"
    }

    @Test
    @Throws(URISyntaxException::class, IOException::class)
    fun testBuildReadmeStreamSBT() {
        val projectSbt = javaClass.getResource("/directory2NoReadme/project5Sbt")
        projectSbt.shouldNotBeNull()
        val path = Path.of(projectSbt.toURI())
        val inputStream = ReadmeNamingParser(
            templateLocation = templatePath,
            fileFilterChain = FileFilterChain.createDefaultChain()
        ).buildReadmeStream(path)
        inputStream.shouldNotBeNull()
        val result = IOUtils.toString(inputStream, defaultCharset())
        result shouldBe "# sbt-project"
    }
}